### DISCUSSION TITLE
Expectations for blueprints and routing

### DISCUSSION THREAD

**Discussion by gg2** [ckan/ckan#6505] [Q&A] [answered]
I'm attempting to update custom extensions to 2.9, and not finished yet, but trying to understand the expectations for how blueprint routes interact.

First off, it looks like before_map and after_map no longer do anything in CKAN 2.9. i.e. All routes must be handled via blueprints. Correct?

Next, if so, and from what I'm seeing, if templates must reference specific blueprints in url_for, etc., then how can one plugin replace or override CKAN's defaults? Is it purely order in which blueprints are registered?
For example, if I create a plugin that modifies authentication, it overrides certain /user/* or /login, /logout, etc routes, and it contains a blueprint named "my_user_auth". The plugin's own templates must call "my_user_auth.login" or "my_user_auth.edit" to get the right results. If other plugins reference CKAN's default "user.edit", how does my authentication plugin overrule CKAN's default?

Maybe this will become clearer shortly as I make better progress upgrading my plugins. But it is very unclear to me from a farther perspective how routes are decided, especially without before_map anymore. I guess the gist is: there seems discontinuity between defining routes in blueprints, but calling specific blueprints (rather than routes) in url_for and similar functions. I guess, I could just hand in hard-coded URLs to url_for...

**Comment by amercader** [comment, marked as answer]
> First off, it looks like before_map and after_map no longer do anything in CKAN 2.9. i.e. All routes must be handled via blueprints. Correct?

Yes, correct

Extension blueprints are registered after core ones, so they will override any core route. Any link that points to this overridden route will point to your custom blueprint, regardless of whether it was constructed with `h.url_for('user.edit', id='test_user')` or `h.url_for('my_user_auth.edit', id='test_user')`. Both these will return `/user/edit/test_user` (which will point to your blueprint).

Does this help?

**Reply by gg2** [reply]
I think so. It sounds like what I missed is that it's not a view/func call directly.

So even if one plugin constructs a route (using url_for) based on one view and view_func, and another plugin does the same based CKAN's default views, what comes out in the end is simply the url, and the url is handled by whichever extension last registered that route. That makes sense.

Thank you!
